src/paginationCalls/pagedApiCaller.ts

```diff
@@ -58,97 +151,131 @@ export class PagedApiCaller implements APICaller {
         callback(
           new GoogleError('Undefined response in pagination method callback.')
         );
         return;
       }
-      const resources = response[self.pageDescriptor.resourceField];
-      const pageToken = response[self.pageDescriptor.responsePageTokenField];
+      const resources = response[resourceFieldName];
+      const pageToken = response[responsePageTokenFieldName];
+      let nextPageRequest = null;
       if (pageToken) {
-        request![self.pageDescriptor.requestPageTokenField] = pageToken;
-        callback(err, resources, request, response);
-      } else {
-        callback(err, resources, null, response);
+        nextPageRequest = Object.assign({}, request);
+        nextPageRequest[requestPageTokenFieldName] = pageToken;
       }
+      callback(err, resources, nextPageRequest, response);
     };
   }
 
+  /**
+   * Adds a special ability to understand pagination-specific fields to the existing gRPC call.
+   * The original gRPC call just calls callback(err, result).
+   * The wrapped one will call callback(err, resources, nextPageRequest, rawResponse) instead.
+   *
+   * @param func gRPC call (normally, a service stub call). The gRPC call is expected to accept four parameters:
+   * request, metadata, call options, and callback.
+   */
   wrap(func: GRPCCall): GRPCCall {
     const self = this;
     return function wrappedCall(argument, metadata, options, callback) {
       return (func as UnaryCall)(
         argument,
         metadata,
         options,
-        self.createActualCallback(argument, callback)
+        self.generateParseResponseCallback(argument, callback)
       );
     };
   }
 
+  /**
+   * Makes it possible to use both callback-based and promise-based calls.
+   * Returns an OngoingCall or OngoingCallPromise object.
+   * Regardless of which one is returned, it always has a `.callback` to call.
+   *
+   * @param settings Call settings. Can only be used to replace Promise with another promise implementation.
+   * @param [callback] Callback to be called, if any.
+   */
   init(settings: ApiCallerSettings, callback?: APICallback) {
     if (callback) {
       return new OngoingCall(callback);
     }
     return new OngoingCallPromise(settings.promise);
   }
 
+  /**
+   * Implements auto-pagination logic.
+   *
+   * @param apiCall A function that performs gRPC request and calls its callback with a response or an error.
+   * It's supposed to be a gRPC service stub function wrapped into several layers of wrappers that make it
+   * accept just two parameters: (request, callback).
+   * @param request A request object that came from the user.
+   * @param settings Call settings. We are interested in `maxResults`, autoPaginate`, `pageToken`, and `pageSize`
+   * (they are all optional).
+   * @param ongoingCall An instance of OngoingCall or OngoingCallPromise that can be used for call cancellation,
+   * and is used to return results to the user.
+   */
   call(
```

Comment:
In the call function, 
here after wrapped:  `apiCall(request, option, metadata, callback)`
but `ResourceCollector` accepts `apiCalls(request, callback)`

Does adding timeout/retry params do the magic? I'm a little bit confused that we wrapped callback functions from gRPC func, and then reconverted back to a simple call(request, callback)

Reply:
Yes, this transition happens in adding retries/timeout.